### PR TITLE
Bit 87 coldkey encryption security

### PR DESCRIPTION
**Pull request by isabella618033**
#2 Update wallet from Ansible Vault to NaCl & argon2i password hashing.

Wallets can be updated with the following scenario 
- `btcli update_wallet --all`
- After running a transfer/stake/unstake/delegate command, keyfile version would be checked and advised to get updated.

(successful update with legacy_wallet_2478, failure in update with legacy_wallet_7807, all other wallets had been updated previously)
image

### CODE REVIEW

File: bittensor/_cli/commands/update_wallet.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            help='''Set true to avoid prompting the user.''',
+            default=False,
+        )
+        bittensor.wallet.add_args( update_wallet_parser )
+        bittensor.subtensor.add_args( update_wallet_parser )
+
+    @staticmethod   
+    def check_config( config: 'bittensor.Config' ):
+        # Ask the user to specify the wallet if the wallet name is not clear. 
+        if config.all == False and config.wallet.get('name') == bittensor.defaults.wallet.name and not config.no_prompt:
```
Comment (camfairchild) [inline bittensor/_cli/commands/update_wallet.py]:
```suggestion
        if config.get('all', d=False) == False and config.wallet.get('name') == bittensor.defaults.wallet.name and not config.no_prompt:
```

File: bittensor/_cli/commands/update_wallet.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+console = bittensor.__console__
+
+class UpdateWalletCommand:
+
+    @staticmethod
+    def run( cli ):
+        '''Check if any of the wallets needs an update.'''
+        config = cli.config.copy()
+        if config.all == True:
```
Comment (camfairchild) [inline bittensor/_cli/commands/update_wallet.py]:
```suggestion
        if config.get('all', d=False) == True:
```

File: tests/unit_tests/bittensor_tests/test_wallet.py [hunk 3]
```diff
@@ -19,6 +19,95 @@
 from unittest.mock import patch, MagicMock
 import pytest
 import bittensor
+from ansible_vault import Vault
+import random
+import getpass
+from rich.prompt import Confirm
+import json
+from bittensor._keyfile.keyfile_impl import encrypt_keyfile_data, decrypt_keyfile_data, keyfile_data_is_encrypted, keyfile_data_is_encrypted_nacl, keyfile_data_is_encrypted_ansible
+class TestWalletUpdate(unittest.TestCase):
```
Comment (eduardogr) [inline tests/unit_tests/bittensor_tests/test_wallet.py]: newlines between imports and class definition?

File: tests/integration_tests/test_cli.py [hunk 4]
```diff
@@ -1854,7 +1871,7 @@ def mock_get_wallet(*args, **kwargs):
             self.assertIsNotNone(mock_console.captured_print)
 
             output_no_syntax = mock_console.remove_rich_syntax(mock_console.captured_print)
-
+
```
Comment (eduardogr) [inline tests/integration_tests/test_cli.py]: extra spaces added

File: bittensor/_keyfile/keyfile_impl.py [hunk 5]
```diff
@@ -143,6 +146,17 @@ def ask_password_to_encrypt() -> str:
...
+def keyfile_data_is_encrypted_nacl( keyfile_data:bytes ) -> bool:
+    """ Returns true if the keyfile data is NaCl encrypted.
+        Args:
+            keyfile_data ( bytes, required ):
+                Bytes to validate
+        Returns:
+            is_nacl (bool):
+                True if data is ansible encrypted.
+    """
+    return keyfile_data[:len('$NACL')] == b'$NACL'
```
Comment (eduardogr) [inline bittensor/_keyfile/keyfile_impl.py:158]: does it make sense to create a variable with the '$NACL' value to not use it twice?

File: bittensor/_keyfile/keyfile_impl.py [hunk 6]
```diff
@@ -227,8 +259,17 @@ def decrypt_keyfile_data(keyfile_data: bytes, password: str = None, coldkey_name
...
         console = bittensor.__console__;             
         with console.status(":key: Decrypting key..."):
+            
+            # NaCl SecretBox decrypt.
+            if keyfile_data_is_encrypted_nacl( keyfile_data ):
+                password = bytes(password, 'utf-8')
+                kdf = pwhash.argon2i.kdf
+                key = kdf(secret.SecretBox.KEY_SIZE, password, NACL_SALT, opslimit=pwhash.argon2i.OPSLIMIT_SENSITIVE, memlimit=pwhash.argon2i.MEMLIMIT_SENSITIVE)
+                box = secret.SecretBox(key)
+                decrypted_keyfile_data = box.decrypt(keyfile_data[len('$NACL'):])
```
Comment (eduardogr) [inline bittensor/_keyfile/keyfile_impl.py:269]: another usage of the value where we could use the variable/constant with '$NACL'

File: tests/unit_tests/bittensor_tests/test_wallet.py [hunk 7]
```diff
@@ -19,6 +19,96 @@
 from unittest.mock import patch, MagicMock
 import pytest
 import bittensor
+from ansible_vault import Vault
+import random
+import getpass
+from rich.prompt import Confirm
+import json
+from bittensor._keyfile.keyfile_impl import encrypt_keyfile_data, decrypt_keyfile_data, keyfile_data_is_encrypted, keyfile_data_is_encrypted_nacl, keyfile_data_is_encrypted_ansible
```
Comment (eduardogr) [inline tests/unit_tests/bittensor_tests/test_wallet.py:27]: we are already importing bittensor, right? Is this needed?

**Review by eduardogr — APPROVED** [review]

**Comment by shibshib** [comment]
This PR is pretty important but stale now, what's the status here @isabella618033 ?

**Comment by camfairchild** [comment]
> This PR is pretty important but stale now, what's the status here @isabella618033 ?

Following up on this. Are we expecting to add this still?

**Comment by ifrit98** [comment]
> This PR is pretty important but stale now, what's the status here @isabella618033 ?

Bump.

**Review by camfairchild — CHANGES_REQUESTED** [review]

**Comment by ifrit98** [comment]
Migrated to: https://github.com/opentensor/bittensor/pull/1497

**Pull request closed without merging**